### PR TITLE
Adding in configurable MaximumRetryAttempts for SQS

### PR DESCRIPTION
**Pull request by ConradKurth**
I added Max retry to SQS because support just became available

Addresses #7012

## How can we verify it

functions:
  compute:
    handler: handler.compute
    events:
      - sqs:
          arn: arn:aws:sqs:region:XXXXXX:stream/foo
          batchSize: 10
          maximumRetryAttempts: 10

## Todos

<details>
<summary>Useful Scripts</summary>

- `npm run test:ci` --> Run all validation checks on proposed changes
- `npm run lint:updated` --> Lint all the updated files
- `npm run lint:fix` --> Automatically fix lint problems (if possible)
- `npm run prettier-check:updated` --> Check if updated files adhere to Prettier config
- `npm run prettify:updated` --> Prettify all the updated files

</details>

- [x] Write and run all tests
- [x] Write documentation
- [x] Enable "Allow edits from maintainers" for this PR
- [ ] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: lib/plugins/aws/package/compile/events/sqs/index.js [hunk 1]
```diff
@@ -28,6 +28,7 @@ class AwsCompileSQSEvents {
             let EventSourceArn;
             let BatchSize = 10;
             let Enabled = true;
+            let MaximumRetryAttempts = 2;
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/sqs/index.js:31]: Let's not set the default (we experience of such change being breaking)

File: docs/providers/aws/events/sqs.md [hunk 2]
```diff
@@ -59,6 +59,28 @@ functions:
...
+
+```yml
+functions:
+  compute:
+    handler: handler.compute
+    events:
+      - sqs:
+          arn: arn:aws:sqs:region:XXXXXX:stream/foo
+          batchSize: 10
+          maximumRetryAttempts: 10
```
Comment (medikoo) [inline docs/providers/aws/events/sqs.md:83]: It'll be great to also have it listed here: https://github.com/serverless/serverless/blob/23bbcea65c3571798435aefc6d6dc9151814cab8/docs/providers/aws/guide/serverless.yml.md

**Comment by medikoo** [comment]
@ConradKurth do you think it really closes #7012 ? Will this mark `MaximumRetryAttempts` being supported in all events?

**Comment by ConradKurth** [comment]
@medikoo No it does not close that ticket, but I was not sure what other issue to link it to. I assume it should reference it, but it does not close it no

**Comment by medikoo** [comment]
> @medikoo No it does not close that ticket,

I've update description them to say _Addresses_ instead of _Closes_ (that will automatically close the #7012)

**Comment by ConradKurth** [comment]
@medikoo Awesome! thanks for the help

**Comment by ConradKurth** [comment]
@medikoo Ready for another look when you have the chance!

File: lib/plugins/aws/package/compile/events/sqs/index.js [hunk 3]
```diff
@@ -146,6 +146,16 @@ class AwsCompileSQSEvents {
...
             };
 
+            if (event.sqs.batchSize) {
+              _.merge(newSQSObject, {
+                [queueLogicalId]: {
+                  Properties: {
+                    MaximumRetryAttempts: event.sqs.maximumRetryAttempts,
+                  },
+                },
+              });
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/sqs/index.js:156]: Let's use plain JS as:

```javascript
newSQSObject[queueLogicalId].Properties.MaximumRetryAttempts = event.sqs.maximumRetryAttempts
```

(idea is to resort to lodash only if we can't achieve something cleanly with plain language constructs)

Comment (ConradKurth) [reply lib/plugins/aws/package/compile/events/sqs/index.js:156]: Ok!

File: lib/plugins/aws/package/compile/events/sqs/index.js [hunk 4]
```diff
@@ -146,6 +146,11 @@ class AwsCompileSQSEvents {
               [queueLogicalId]: JSON.parse(sqsTemplate),
             };
 
+            if (event.sqs.batchSize) {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/sqs/index.js:149]: Just to be sure, why do we check on `event.sqs.batchSize` here and not `event.sqs.maximumRetryAttempts` ?

If it's intentional it'll be good to add some comment clarifying such logic route

Comment (ConradKurth) [reply lib/plugins/aws/package/compile/events/sqs/index.js]: haha nice! some good copy pasta right there. Fixed!

**Comment by mnapoli** [comment]
That's really interesting, can anyone explain how this is different from setting up `maxReceiveCount` on the SQS queue?

For example:

```yaml
functions:
    worker:
        handler: handler.handler
        events:
            -   sqs:
                    arn:
                        Fn::GetAtt: [ Queue, Arn ]

resources:
    Resources:
        Queue:
            Type: AWS::SQS::Queue
            Properties:
                RedrivePolicy:
                    deadLetterTargetArn: ...
                    # Jobs will be retried 5 times
                    # The number needs to be at least 5 per https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html
                    maxReceiveCount: 5
```

File: docs/providers/aws/events/sqs.md (hunk 2 above)
Comment (sanrodari) [inline docs/providers/aws/events/sqs.md:83]: I think the only valid values are: 0, 1, 2
@ConradKurth, @medikoo

Comment (medikoo) [reply docs/providers/aws/events/sqs.md:83]: @sanrodari Documentation states any number in `0...10000` range is accepted: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html#cfn-lambda-eventsourcemapping-maximumretryattempts

Comment (sanrodari) [reply docs/providers/aws/events/sqs.md:83]: Got it, thanks

**Comment by demian85** [comment]
Can someone please confirm if this setting also works for async invocations without event sources? I do not have an SQS event, I just have a plain lambda that needs a retry count of 0 by default. How can I achieve that?

**Comment by medikoo** [comment]
> Can someone please confirm if this setting also works for async invocations without event sources? 

It should be possible shortly. See https://github.com/serverless/serverless/issues/7678#issuecomment-626188047

**Comment by MicMonen** [comment]
> That's really interesting, can anyone explain how this is different from setting up `maxReceiveCount` on the SQS queue?
> 
> For example:
> 
> ```yaml
> functions:
>     worker:
>         handler: handler.handler
>         events:
>             -   sqs:
>                     arn:
>                         Fn::GetAtt: [ Queue, Arn ]
> 
> resources:
>     Resources:
>         Queue:
>             Type: AWS::SQS::Queue
>             Properties:
>                 RedrivePolicy:
>                     deadLetterTargetArn: ...
>                     # Jobs will be retried 5 times
>                     # The number needs to be at least 5 per https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html
>                     maxReceiveCount: 5
> ```

I am curious also whats the difference.

**Comment by a428tm** [comment]
Hi, I could be doing something wrong, but if i understood it correctly, I should be able to set maxRetry? https://www.serverless.com/framework/docs/providers/aws/events/sqs/

When I do this, 
`functions:
  abcd:
    handler: abcd.handler
    description: POST /abcd
    events:
      - http:
          path: abcd
          method: post
          cors: true
          private: true # for API keys
      - sqs:
          arn:
            Fn::GetAtt:
              - InvokeRekognitionQueue
              - Arn
          maximumRetryAttempts: 2
      - sqs:
          arn:
            Fn::ImportValue: ${self:custom.stage}-DDBStreamQueueArn
          maximumRetryAttempts: 2`

I get this error - 

>   Serverless Error ---------------------------------------
 
  An error occurred: devDDBStreamQueueArn - MaxiumumRetryAttempts isn't supported for this event source type. (Service: AWSLambda; Status Code: 400; Error Code: InvalidParameterValueException; Request ID: a25ac6c2-41d9-4f73-85fb-700436c558c7).

Am I missing something here...?

**Comment by mnapoli** [comment]
Is it just me or is `maximumRetryAttempts` just for DynamoDB and Kinesis?

That would explain my question: why `maximumRetryAttempts` when we already have `maxReceiveCount`?

Oh wait I just found this:

CleanShot 2020-06-09 at 09 55 09

Source: https://github.com/awsdocs/aws-lambda-developer-guide/blob/master/doc_source/API_CreateEventSourceMapping.md

It seems to support that `MaximumRetryAttempts` is not available for SQS. Should this PR be reverted?

**Comment by medikoo** [comment]
@mnapoli great thanks for pointing that here. We have also that reported here: https://github.com/serverless/serverless/issues/7762

It totally should be reverted. I will follow with it now

**Comment by demian85** [comment]
How can you revert a feature that I'm already using in production? Are you
serious? It should be implemented where missing instead.

**Comment by demian85** [comment]
Ops, sorry. I thought you were going to revert the option for kinesis as
well. My bad.

**Comment by jstallm** [comment]
What is the interplay between a lambdas "maximum_retry_attempts" property and an SQS "maxReceiveCount"?

**Comment by mnapoli** [comment]
@jstallm there isn't, `maxReceiveCount` doesn't apply with SQS.

See also https://github.com/awsdocs/aws-cloudformation-user-guide/pull/715

**Comment by jstallm** [comment]
@mnapoli , are you 100% sure about that? 
Here is verbatim directly from documentation 
“ If your function returns an error, or can't be invoked because it's at maximum concurrency, processing might succeed with additional attempts. To give messages a better chance to be processed before sending them to the dead-letter queue, set the maxReceiveCount on the source queue's redrive policy to at least 5.”

https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html#events-sqs-queueconfig

**Comment by mnapoli** [comment]
@jstallm woops my bad, I should have written:

> `maximumRetryAttempts` doesn't apply with SQS.

- `maximumRetryAttempts` is a Lambda setting for asynchronous invocations
- `maxReceiveCount` is a SQS setting

SQS does not invoke Lambda asynchronously, it invokes it synchronously. That's why `maximumRetryAttempts` is useless for SQS.

**Comment by jstallm** [comment]
@mnapoli , thank you for this information. Is this fact documented anywhere today ?

**Comment by mnapoli** [comment]
I opened https://github.com/awsdocs/aws-cloudformation-user-guide/pull/715 to try to clarify the AWS documentation but that isn't merged yet.

Also see [my previous comment](https://github.com/serverless/serverless/pull/7620#issuecomment-641104677).

If you are looking for more, I'll shamelessly plug https://serverless-visually-explained.com where I show this in details (e.g. the difference between Lambda's async invocation and how SQS invokes Lambda).

**Comment by medikoo** [comment]
@mnapoli that looks great, thanks for sharing!

**Pull request merged**